src/ledger/LedgerTxn.cpp

```diff
@@ -1874,27 +2065,28 @@ findIncludedOffer(std::list<LedgerEntry>::const_iterator iter,
 }
 
 std::shared_ptr<LedgerEntry const>
-LedgerTxnRoot::Impl::getBestOffer(Asset const& buying, Asset const& selling,
-                                  std::unordered_set<LedgerKey>& exclude)
+LedgerTxnRoot::Impl::getBestOffer(LedgerEntry const& worseThan)
 {
+    auto const& oe = worseThan.data.offer();
+
     // Note: Elements of mBestOffersCache are properly sorted lists of the best
     // offers for a certain asset pair. This function maintaints the invariant
     // that the lists of best offers remain properly sorted. The sort order is
     // that determined by loadBestOffers and isBetterOffer (both induce the same
     // order).
-    auto cached = getFromBestOffersCache(buying, selling);
+    auto cached = getFromBestOffersCache(oe.buying, oe.selling);
     auto& offers = cached->bestOffers;
 
-    auto res = findIncludedOffer(offers.cbegin(), offers.cend(), exclude);
+    auto res = findIncludedOffer(offers.cbegin(), offers.cend(), worseThan);
 
     size_t const BATCH_SIZE = 5;
     while (!res && !cached->allLoaded)
     {
         std::list<LedgerEntry>::const_iterator newOfferIter;
         try
         {
-            newOfferIter = loadBestOffers(offers, buying, selling, BATCH_SIZE,
-                                          offers.size());
+            newOfferIter = loadBestOffers(offers, oe.buying, oe.selling,
```

Comment:
I've already improved the SQL query considerably in a future commit, taking advantage of the index and the knowledge of `worseThan`. But it is built on top of this PR, so I'll publish it once we've worked out the issues in this one.

Reply:
cool. I realized later that we can't actually use `worseThan` in this version as it may cause unwanted gaps in `offers`. I'll watch for the change in the later PR that makes changes to this area.